File: libgomp/testsuite/libgomp.fortran/target-map-iterators-1.f90
! { dg-do run }

! Test transfer of dynamically-allocated arrays to target using map
! iterators.

program test
  implicit none

  integer, parameter :: DIM1 = 8
  integer, parameter :: DIM2 = 15

  type :: array_ptr
    integer, pointer :: arr(:)
  end type

  type (array_ptr) :: x(DIM1)
  integer :: expected, sum, i, j

  expected = mkarray ()

  !$omp target map(iterator(i=1:DIM1), to: x(i)%arr(:)) map(from: sum)
    sum = 0
    do i = 1, DIM1
      do j = 1, DIM2
	sum = sum + x(i)%arr(j)
      end do
    end do
  !$omp end target

  if (sum .ne. expected) stop 1
contains
  integer function mkarray ()
    integer :: exp = 0

    do i = 1, DIM1
      allocate (x(i)%arr(DIM2))
      do j = 1, DIM2
        x(i)%arr(j) = i * j
	exp = exp + x(i)%arr(j)
      end do
    end do

    mkarray = exp
  end function
end program

File: libgomp/testsuite/libgomp.c-c++-common/target-map-iterators-1.c
/* { dg-do run } */
/* { dg-require-effective-target offload_device_nonshared_as } */

/* Test transfer of dynamically-allocated arrays to target using map
   iterators.  */

#include <stdlib.h>

#define DIM1 8
#define DIM2 15

int mkarray (int *x[])
{
  int expected = 0;

  for (int i = 0; i < DIM1; i++)
    {
      x[i] = (int *) malloc (DIM2 * sizeof (int));
      for (int j = 0; j < DIM2; j++)
	{
	  x[i][j] = rand ();
	  expected += x[i][j];
	}
    }

  return expected;
}

int main (void)
{
  int *x[DIM1];
  int y;

  int expected = mkarray (x);

  #pragma omp target enter data map(to: x)
  #pragma omp target map(iterator(i=0:DIM1), to: x[i][:DIM2]) \
		     map(from: y)
    {
      y = 0;
      for (int i = 0; i < DIM1; i++)
	for (int j = 0; j < DIM2; j++)
	  y += x[i][j];
    }

  return y - expected;
}
